lms/services/oauth_http.py

```diff
@@ -0,0 +1,52 @@
+from lms.services import HTTPError
+
+
+class OAuthHTTPService:
+    """Send OAuth 2.0 requests and return the responses."""
+
+    def __init__(self, http_service, oauth2_token_service):
+        self._http_service = http_service
+        self._oauth2_token_service = oauth2_token_service
+
+    def get(self, *args, **kwargs):
+        return self.request("GET", *args, **kwargs)
+
+    def put(self, *args, **kwargs):
+        return self.request("PUT", *args, **kwargs)
+
+    def post(self, *args, **kwargs):
+        return self.request("POST", *args, **kwargs)
+
+    def patch(self, *args, **kwargs):
+        return self.request("PATCH", *args, **kwargs)
+
+    def delete(self, *args, **kwargs):
+        return self.request("DELETE", *args, **kwargs)
+
```

Comment:
```suggestion
class OAuthHTTPService(HTTPService):
    """Send OAuth 2.0 requests and return the responses."""

    def __init__(self,  oauth2_token_service):
        super().__init__()
        self._oauth2_token_service = oauth2_token_service
```

In the same way that I think the `HTTPService` is a `Session` rather than having one, I think the `OAuthHTTPService` is an `HTTPService`. We are repeatedly implementing a lot of stub methods with no functionality in them which replicate those in `HTTPService` which replicate those in `Session`. I think this (along with the name) is a strong hint for inheritance instead of composition.

Reply:
> I think Session -> HTTPService -> OAuthHTTPService is a fairly clear cut case of inheritance over composition. There are 11 methods which basically just proxy up to the parent in the current design which could be deleted. I think this is about a strong a signal for inheritance as you're going to get. The name as well strongly hints what's happening here. For example these could easily be Session -> HTTPSession -> OAuthHTTPSession.

I think I'd like to stick with the composition-based design here. The intention with `HTTPService` (and now also `OAuthHTTPService`) is to provide a **simpler** interface than the one that `requests` provides, as well as to embody best practices for using `requests` correctly (using a session, handling errors). So I _don't_ think that `HTTPService` "is a" `requests.Session`. It's a service that wraps `requests` and provides a simpler interface.

I don't actually want `HTTPService` to expose all of `requests.Session`'s attributes (`close()`, `mount()`, etc) as its own public interface.

This does have the downside of creating a certain amount of boilerplate because `HTTPService` needs `get()`, `put()`, `post()`, `patch()` and `delete()` methods that just forward to its `request()` method, and `OAuthHTTPService` needs the same (as does the Blackboard's `BasicClient`). (They actually don't really need these. You could just call `request("METHOD", ...)`. But they're nice to have.) If the duplicated boiler is the problem I think we could use a mixin class to reduce that without losing the benefits of composition / taking on the downsides of inheritance:

```python
class HTTPMethodMixin:
    """Add HTTP verb methods (get(), post(), etc) that forward to self.request()."""

    def get(self, *args, **kwargs):
        return self.request("GET", *args, **kwargs)

    def put(self, *args, **kwargs):
        return self.request("PUT", *args, **kwargs)

    def post(self, *args, **kwargs):
        return self.request("POST", *args, **kwargs)

    def patch(self, *args, **kwargs):
        return self.request("PATCH", *args, **kwargs)

    def delete(self, *args, **kwargs):
        return self.request("DELETE", *args, **kwargs)
```

Anyway, this is why I prefer composition over inheritance:

### I don't think the duplicated forwarding methods are a big problem

I think duplicated boilerplate code is a minor sin (and possibly amenable to the mixin-based solution I suggested above) but I don't think it's a big problem.

On the other hand I **do** think that the sins of inheritance are a big problem: tightly coupling `HTTPService` to `requests.Session` (and `OAuthHTTPService` and `BasicClient` to `HTTPService`); getting in the way of isolated unit testing (it will no longer be easy to test `OAuthHTTPService` and `BasicClient` in isolation with a mock `HTTPService`); harming code readability; loss of flexibility. I think [this post](https://hynek.me/articles/python-subclassing-redux/) covers it well (particularly the **Type 1: Code Sharing** section).

### I don't expect the amount of boiler to grow

`HTTPService` originally just had `request()` and Marcos requested adding `get()` and pals for convenience. I don't expect to want to add many other `requests.Session` attributes to `HTTPService`'s public interface so I don't expect the amount of boiler here to grow much. (At some point if you have to write a really large number of forwarding methods this can tip the balance towards inheritance but I don't think we're hitting that point here.)

### This is just Python's way of doing composition

The boiler is just sort of par-for-the-course with composition in Python. Other languages have more concise ways to declare when a class forwards methods to a composed object. Ruby has this for example. There's even a third-party Python lib called [forwardable](https://github.com/5long/forwardable) that implements something like this in Python:

```python
@forwardable()
class HTTPService:
    def_delegators("_session", "get, put, post, patch, delete")

    def __init__(self, ..._):
        self._session = requests.Session()
        ...

    def request(self, ...):
        ...
```

I don't think we should use forwardable. The point I'm trying to make is that boilerplatey forwarding methods are just how you do it in Python so we shouldn't consider them too offensive.

### They're not actually forwarding methods

The `get()`, `put()`, `post()`, `patch()` and `delete()` **don't** actually forward to `requests.Session.get()`, `requests.Session.put()`, `requests.Session.post()`, `requests.Session.patch()` and `requests.Session.delete()`. They forward to `HTTPService.request()` which wraps `requests.Session.request()` and adds custom behaviour. 

It may be that in the `requests` source code `requests.Session.get()` etc do just forward to `self.request()` in which case sub-classing `requests.Session` and just overriding `request()` would work. But this would mean that `HTTPService` now implicitly depends on private implementation details of `requests.Session` and will break if `requests.Session` changes its internal implementation details.

Some classes are explicitly designed for subclassing and provide methods that you're supposed to override and guarantees about how and when those methods will be called. AFAIK `requests.Session.request()` isn't intended to be used in this way, it might just happen to work by accident.

Reply:
> I'm not sure OAuthHTTPService is doing all it should. I'd really like to see the non-Blackboard parts get_token() coming from the service too. This should be a standardised thing for the most part, and I think you'd only need to pass in the base URL to make this work. I think it's a good sign that OAuthHTTPService has a well defined boundary when we don't need the token service outside of it.

I think you're very likely right about this one but I'd like to hold off until we have more than one user of `OAuthHTTPService`. Then we'll be able to see what the different implementations of `get_token()` have in common and where they differ. I thought I'd said this somewhere, but I'm just holding off on this one until we have more cases.

Reply:
Added a PR to split out `OAuthHTTPService`, sticking with the composition-based design: https://github.com/hypothesis/lms/pull/2985

Reply:
> > I'm not sure OAuthHTTPService is doing all it should. I'd really like to see the non-Blackboard parts get_token() coming from the service too. This should be a standardised thing for the most part, and I think you'd only need to pass in the base URL to make this work. I think it's a good sign that OAuthHTTPService has a well defined boundary when we don't need the token service outside of it.
> 
> I think you're very likely right about this one but I'd like to hold off until we have more than one user of `OAuthHTTPService`. Then we'll be able to see what the different implementations of `get_token()` have in common and where they differ. I thought I'd said this somewhere, but I'm just holding off on this one until we have more cases.

I ended up writing generic `access_token_request()` and `refresh_token_request()` methods anyway even though they're only used for the Blackboard API so far: https://github.com/hypothesis/lms/pull/2987, https://github.com/hypothesis/lms/pull/2988. These won't necessarily work for other servers besides Blackboard yet but we can generalize them further as we need to when we come to it.

I ended up adding these to `OAuth2TokenService` not `OAuthHTTPService` because `OAuthHTTPService.request()` is going to contain higher-level "send with refresh and retry" logic that I want to unit test with a mock `refresh_token_request()` method.